Add Apple Watch companion app

jasonzurita opened the pull request:
Edit: Now ready to be merged!

---

The initial commit here is a good first start, but is not quite ready to be merged.

Instead of merging this into master, this PR will be opened and subsequent PRs will be made to this branch to keep the PRs small and digestible. When the full watch feature is ready, this PR will be merged with everything!

jasonzurita committed:
feat(watch): add stock apple watch extension

Slightly modified from the Xcode provided watch
template to clean up thing like comments.

Start of the watch companion (maybe stand alone)
watch app!!!

jasonzurita committed:
chore: configure watch target to run on device

The build number and version number need
consistency between targets. Version number
still needs to be manually set, but the build number
now uses the same _commit count_ run script as the
main app.

Also, remove comment from main app run script.

jasonzurita committed:
feat: implement first working watch version

There is still a lot to do, but this is a great first
start! (see `HomeView` for to do list)

FlippingLeaf commented on BabyPatterns/FeedingVC.swift:
```diff
@@ -70,6 +70,23 @@ final class FeedingVC: UIViewController {
+        notificationToken = center.addObserver(forName: K.Notifications.updateFeedingsUI,
+                                               object: nil,
+                                               queue: nil)
+        { _ in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/BabyPatterns/FeedingVC.swift#L78"></span>Opening braces should be preceded by a single space and on the same line as the declaration.<br />
<code>opening_brace</code> <code>FeedingVC.swift:78</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on BabyPatterns/FeedingsVM.swift:
```diff
@@ -1,11 +1,74 @@
 
-class FeedingsVM: Loggable {
+extension FeedingsVM: WCSessionDelegate {
+    public func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/BabyPatterns/FeedingsVM.swift#L7"></span>Line should be 120 characters or less: currently 131 characters<br />
<code>line_length</code> <code>FeedingsVM.swift:7</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on BabyPatterns/FeedingsVM.swift:
```diff
@@ -1,11 +1,74 @@
+    public func sessionDidDeactivate(_ session: WCSession) {
+        print("session did deactivate")
+    }
+    public func session(_ session: WCSession, didReceiveMessage message: [String : Any]) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/BabyPatterns/FeedingsVM.swift#L18"></span>Colons should be next to the identifier when specifying a type and next to the key in dictionary literals.<br />
<code>colon</code> <code>FeedingsVM.swift:18</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on BabyPatterns/FeedingsVM.swift:
```diff
@@ -1,11 +1,74 @@
+    public func session(_ session: WCSession, didReceiveMessage message: [String : Any]) {
+        print("Received message!")
+    }
+    public func session(_ session: WCSession, didReceiveMessage message: [String : Any], replyHandler: @escaping ([String : Any]) -> Void) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/BabyPatterns/FeedingsVM.swift#L21"></span>Line should be 120 characters or less: currently 140 characters<br />
<code>line_length</code> <code>FeedingsVM.swift:21</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/BabyPatterns/FeedingsVM.swift#L21"></span>Colons should be next to the identifier when specifying a type and next to the key in dictionary literals.<br />
<code>colon</code> <code>FeedingsVM.swift:21</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/AddFeedingView.swift:
```diff
@@ -0,0 +1,106 @@
+                        // TODO: These numbers are not exact, figure this out better
+                        .offset(x: self.intent.isAdding ? -45 : -metrics.size.width * 0.5 - 50)
+                        .gesture(TapGesture().onEnded {
+                            WCSession.default.sendMessage(["feedingType": "\(self.intent.rawValue)", "side": "left"], replyHandler: { reply in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/0d2b465163fcdd25369a82af0b5631fc627ea383/Watch Extension/AddFeedingView.swift#L77"></span>Line should be 120 characters or less: currently 142 characters<br />
<code>line_length</code> <code>AddFeedingView.swift:77</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/AddFeedingView.swift:
```diff
@@ -0,0 +1,106 @@
+                        .gesture(TapGesture().onEnded {
+                            WCSession.default.sendMessage(["feedingType": "\(self.intent.rawValue)", "side": "left"], replyHandler: { reply in
+                                print(reply)
+                            }) { e in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/0d2b465163fcdd25369a82af0b5631fc627ea383/Watch Extension/AddFeedingView.swift#L79"></span>Trailing closure syntax should not be used when passing more than one closure argument.<br />
<code>multiple_closures_with_trailing_closure</code> <code>AddFeedingView.swift:79</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/AddFeedingView.swift:
```diff
@@ -0,0 +1,106 @@
+                        .font(.system(size: 55))
+                        .offset(x: self.intent.isAdding ? 45 : metrics.size.width * 0.5 + 50)
+                        .gesture(TapGesture().onEnded {
+                            WCSession.default.sendMessage(["feedingType": "\(self.intent.rawValue)", "side": "right"], replyHandler: { reply in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/0d2b465163fcdd25369a82af0b5631fc627ea383/Watch Extension/AddFeedingView.swift#L89"></span>Line should be 120 characters or less: currently 143 characters<br />
<code>line_length</code> <code>AddFeedingView.swift:89</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/AddFeedingView.swift:
```diff
@@ -0,0 +1,106 @@
+                        .gesture(TapGesture().onEnded {
+                            WCSession.default.sendMessage(["feedingType": "\(self.intent.rawValue)", "side": "right"], replyHandler: { reply in
+                                print(reply)
+                            }) { e in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/0d2b465163fcdd25369a82af0b5631fc627ea383/Watch Extension/AddFeedingView.swift#L91"></span>Trailing closure syntax should not be used when passing more than one closure argument.<br />
<code>multiple_closures_with_trailing_closure</code> <code>AddFeedingView.swift:91</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    func applicationDidBecomeActive() {
+        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/ExtensionDelegate.swift#L16"></span>Line should be 120 characters or less: currently 194 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:16</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    func handle(_ backgroundTasks: Set<WKRefreshBackgroundTask>) {
+        // Sent when the system needs to launch the application in the background to process tasks. Tasks arrive in a set, so loop through and process each one.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/ExtensionDelegate.swift#L25"></span>Line should be 120 characters or less: currently 160 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:25</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+                backgroundTask.setTaskCompletedWithSnapshot(false)
+            case let snapshotTask as WKSnapshotRefreshBackgroundTask:
+                // Snapshot tasks have a unique completion call, make sure to set your expiration date
+                snapshotTask.setTaskCompleted(restoredDefaultState: true, estimatedSnapshotExpiration: Date.distantFuture, userInfo: nil)
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/ExtensionDelegate.swift#L34"></span>Line should be 120 characters or less: currently 137 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:34</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+}
+
+extension ExtensionDelegate: WCSessionDelegate {
+    func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/ExtensionDelegate.swift#L56"></span>Line should be 120 characters or less: currently 124 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:56</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/HomeView.swift:
```diff
@@ -0,0 +1,45 @@
+    var body: some View {
+        VStack {
+            // TODO: fill this in with active feedings
+            List((1...10), id: \.self) { num in
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/HomeView.swift#L21"></span>Unused parameter “num” in a closure should be replaced with _.<br />
<code>unused_closure_parameter</code> <code>HomeView.swift:21</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/AddFeedingView.swift:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+struct ContentView_Previews: PreviewProvider {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/0d2b465163fcdd25369a82af0b5631fc627ea383/Watch Extension/AddFeedingView.swift#L102"></span>Type name should only contain alphanumeric characters: ‘ContentView_Previews’<br />
<code>type_name</code> <code>AddFeedingView.swift:102</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    func applicationWillResignActive() {
+        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/ExtensionDelegate.swift#L20"></span>Line should be 120 characters or less: currently 285 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:20</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/HomeView.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+struct HomeView_Previews: PreviewProvider {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/c49e444d61917df61aa42d9cccb11a46857e6fb9/Watch Extension/HomeView.swift#L41"></span>Type name should only contain alphanumeric characters: ‘HomeView_Previews’<br />
<code>type_name</code> <code>HomeView.swift:41</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented:
<!--
  0 Errors
  2 Warnings: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

FlippingLeaf commented on BabyPatterns/FeedingsVM.swift:
```diff
@@ -1,11 +1,76 @@
+        print("Received message!")
+    }
+
+    public func session(_: WCSession, didReceiveMessage message: [String: Any], replyHandler: @escaping ([String: Any]) -> Void) {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/9a1d22d8ad71ad19efd2b877aa3b663fe9f8e3a5/BabyPatterns/FeedingsVM.swift#L23"></span>Line should be 120 characters or less: currently 130 characters<br />
<code>line_length</code> <code>FeedingsVM.swift:23</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    func applicationDidBecomeActive() {
+        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/9a1d22d8ad71ad19efd2b877aa3b663fe9f8e3a5/Watch Extension/ExtensionDelegate.swift#L15"></span>Line should be 120 characters or less: currently 194 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:15</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    func handle(_ backgroundTasks: Set<WKRefreshBackgroundTask>) {
+        // Sent when the system needs to launch the application in the background to process tasks. Tasks arrive in a set, so loop through and process each one.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/9a1d22d8ad71ad19efd2b877aa3b663fe9f8e3a5/Watch Extension/ExtensionDelegate.swift#L24"></span>Line should be 120 characters or less: currently 160 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:24</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,58 @@
+                backgroundTask.setTaskCompletedWithSnapshot(false)
+            case let snapshotTask as WKSnapshotRefreshBackgroundTask:
+                // Snapshot tasks have a unique completion call, make sure to set your expiration date
+                snapshotTask.setTaskCompleted(restoredDefaultState: true, estimatedSnapshotExpiration: Date.distantFuture, userInfo: nil)
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/9a1d22d8ad71ad19efd2b877aa3b663fe9f8e3a5/Watch Extension/ExtensionDelegate.swift#L33"></span>Line should be 120 characters or less: currently 137 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:33</code></td>
    </tr>
  </tbody>
</table>

FlippingLeaf commented on Watch Extension/ExtensionDelegate.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    func applicationWillResignActive() {
+        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/9a1d22d8ad71ad19efd2b877aa3b663fe9f8e3a5/Watch Extension/ExtensionDelegate.swift#L19"></span>Line should be 120 characters or less: currently 285 characters<br />
<code>line_length</code> <code>ExtensionDelegate.swift:19</code></td>
    </tr>
  </tbody>
</table>

jasonzurita committed:
chore: move watch to use common feeding enums

This introduces a new SPM package called `Common`,
which will be the cross platform library for iOS / watchOS.

This didn't yet make use of the other enums or decide if all
of them make sense in this new module, so more to come.

jasonzurita committed:
feat: implement codable watch/ios app communication

This will simplify the communication between the two,
while improving the robustness of the communication
(get away from strings!)

FlippingLeaf commented on Watch Extension/HomeView.swift:
```diff
@@ -0,0 +1,47 @@
+    }
+}
+
+struct HomeView_Previews: PreviewProvider {
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/jasonzurita/BabyPatterns/blob/4ac60e68c7da14f5f694665403728bcfc55165a4/Watch Extension/HomeView.swift#L43"></span>Type name should only contain alphanumeric characters: ‘HomeView_Previews’<br />
<code>type_name</code> <code>HomeView.swift:43</code></td>
    </tr>
  </tbody>
</table>

jasonzurita committed:
Merge pull request #4 from jasonzurita/feature/improve-communication

Improve watch <> iOS communication

jasonzurita committed:
chore: remove completed to do

jasonzurita committed:
feat: add unidirectional flow state management

This now allows the home screen to react to feedings
being added.

jasonzurita committed:
Merge pull request #5 from jasonzurita/feature/unidirectional-data-flow

Add unidirectional flow state management

jasonzurita committed:
feat: implement watch pause and resume

This further continues the work from the unidirectional
state management. There is some duplicated code / state
between the watch and the iOS app, but that is okay tech debt
for now.

jasonzurita committed:
chore: update gitignore

no need to track this

jasonzurita committed:
chore: use `isEmpty` over `.count > 0`

Using `isEmpty` is more Swifty and more
performant

jasonzurita committed:
Merge pull request #6 from jasonzurita/feature/implement-pause-and-stop

Implement watch pause and resume

jasonzurita committed:
feat: add Apple Watch icons

jasonzurita committed:
Merge pull request #7 from jasonzurita/feature/watch-icons

Add Apple Watch icons

jasonzurita committed:
feat: better style in-progress UI

This will now show what type of feeding and side

jasonzurita committed:
feat: move app environment to common package

This will allow watch use of the app environment's
timer!

Platform versions were added to Package.swift to let the
compiler know that the timer function used in app
environment is available.

jasonzurita committed:
feat: implement a _real_ in-progress view

- Functional counting timer!
- Better order for the in-progress feedings with
the most recent one added at the top of the list

This works based on having a global clock that
all timers _tick_ off of. This seems like a better pattern
than having each in-progress feeding manage their own
timer like the way the iOS app works. Probably something to
see how it works and carry over to the iOS app later :).

The double extension for time formatting was put in the
common package to support cross platform use.

jasonzurita committed:
fix: correct `.isEmpty` checks

The previous improvement to using
`.isEmpty` was wrong and should have been
negated like this. Test would have caught this...

jasonzurita committed:
chore: update todos

jasonzurita committed:
Merge pull request #8 from jasonzurita/feature/feeding-in-progress-ui

Improve feeding in-progress UI (timer + information)

jasonzurita committed:
fix: force passing dummy test to let SwiftUI canvas run

Weird that SwiftUI seems to run tests or compile them
when making the live canvas

jasonzurita committed:
chore: move watch communication to dispatch vc

This will help in communicating logged in / out since
the FeedingVM has no idea about logged in state

jasonzurita committed:
feat: manage logged in/out on watch via context

The main app will update the context to let the watch
know if the user has been logged in or out. The watch then
updates the shown UI accordingly!

jasonzurita committed:
Merge pull request #9 from jasonzurita/feature/manage-logging-in_out

Implement management for logging in out

jasonzurita committed:
feat: surface some errors in communication

Alert error will be shown when reachability is
false or if we can't encode the info to send.

Handling the error completion handler when
sending a message is something for the future
since that is a bit trickier.

jasonzurita committed:
chore: disable one off linter rule

function body length is okay here
since this is a switch statement with clear
branching logic as a result

jasonzurita committed:
Merge pull request #10 from jasonzurita/feature/improve-communication

Surface some errors in communication

jasonzurita committed:
chore: improve UI a bit

The counter's font needed to be smaller
to accommodate the larger width from some
of the time values like 00:00:55

jasonzurita committed:
chore: remove unnecessary dispatch

all state changes via the send method
are already on the main thread.

jasonzurita committed:
chore: update fastlane

jasonzurita committed:
feat: make in-progress feeding available

to be used the the timer label to move use
the model as the single source of truth

jasonzurita committed:
feat: move timer label to use root source of truth

The timer labels now depend on the in-progress
feeding model to show the time, count up, pause, and stop.
This is more reliable in that we are no longer duplicating state
(which interestingly enough used to be the case, but for
some reason past me had the "bright idea" to isolate the
timer label from the single source of truth...oh well :)).

Impact:
- Improves reaction to changes to in-progress feedings from the
Apple Watch app
- Fixes a bug where the paused time was not correctly accounted for
in the timer UI even though the model was correct

Also, update tests.

jasonzurita committed:
Merge pull request #11 from jasonzurita/feature/improve-ios-watchos-sync

Improve main app timer behavior

jasonzurita committed:
chore: update todos

jasonzurita committed:
feat: add debug assertions

This will help in figuring out the reason
why the main app sometimes doesn't get
when the watch changes a feeding

jasonzurita committed:
feat: show main app fyi dialog when watch ends feeding

So that the user sees the same experience on the main
app when a user ends a feeding on the watch

jasonzurita committed:
chore: rename toast to fyi dialog for clarity

A toast should be something that appears and
disappears like this `FyiDialog`, but slides up from
the bottom. The use of the name _Toast_ was
incorrect and inconsistent with the understanding
of the term _Toast_ in this context.

jasonzurita committed:
feat: add saved fyi dialog to watch app

To both:
- provide consistency with the main app
- better UX

jasonzurita committed:
feat: cut bottle feeding from v1 scope

- Simplifies v1 a bit and isn't as important as
the other types of feeding
- Polish up adding new feeding UI a bit
for clarity

jasonzurita committed:
Merge pull request #12 from jasonzurita/feature/feedback-when-watch-ends-feeding

Feedback when watch ends feeding

jasonzurita committed:
Merge pull request #13 from jasonzurita/feature/bottle-feeding

Cut bottle feeding from v1 scope

jasonzurita committed:
feat: prevent starting duplicate feedings

The main app is currently not set up to allow
starting duplicate feedings of the same type.
This prevents that case.

jasonzurita committed:
Merge pull request #14 from jasonzurita/feature/prevent-starting-duplicate-feedings

Prevent starting duplicate feedings

jasonzurita committed:
feat: account for paused feeding

The watch will now properly account for
paused time after a period of being paused.

Right now the UI just says "PAUSED".
In the future we can build out a blinking
paused timer label like the main app, but this
is okay for now.

Heavily based off of the main app's
implementation of the feeding struct. The
two different feeding structs (watch and iOS)
should be combined in the future.

I am not too crazy about duplicating the state
of in-progress feedings here, but due to the
difficulty in watch<>iOS app communication,
this is okay for now.

A resume action was formally built in too.

jasonzurita committed:
chore: separate out feeding struct for clarity

This will eventually be combined with the main
app's Feeding struct, but put here now for better
clarity.

jasonzurita committed:
feat: split up the app reducers

This is the first step in isolating each reducer,
which will enable things like better modularity,
testability, clarity, higher order reducers, etc.!

The piece that isn't great here is that each
smaller reducer takes in the full app state and
app action. Therefore there is a need to default
the switch statement (no future help from the compiler),
and each reducer has too much visibility into the apps
state and capabilities.

jasonzurita committed:
Merge pull request #15 from jasonzurita/feature/account-for-pause

Account for paused feeding

jasonzurita committed:
feat: implement pullback for AppState

This moves the reducers, where applicable,
over to operate over minimum required state
such as integers vs the whole app state.

These reducers now don't have knowledge
about the full AppState and therefore improves
reasoning about what the reducer does and prevents
future bugs due to a reducer doing more than it should
with the full AppState.

Note: feedingReducer will not be able to be simplified until
pullback is defined for AppAction too.

This is the 2nd step in enabling higher order reducers:
https://www.pointfree.co/episodes/ep71-composable-state-management-higher-order-reducers

jasonzurita committed:
feat: implement pullback for actions

This now sets the stage to add higher
order reducers!

The reducers operate on a subset of the
actions, which is less bug prone and more
straight forward (easier to understand).

jasonzurita committed:
Merge pull request #16 from jasonzurita/feature/improve-reducer-and-actions

Improve reducer and actions

jasonzurita committed:
chore: rename file for clarity

unidirectional flow is a better name for this
architecture and the functions that enable it

jasonzurita committed:
chore: split up the app state file for clarity

This file was overloaded with implementations.
Now that there are reducers that operate over
limited state and action, this pulls them out into
their own dedicated file.

Better clarity and easier to follow too!

jasonzurita committed:
feat: add higher order feeding comm reducer

The higher order feeding communication reducer
better encapsulates the communication with the parent
app!

Eventually (and as noted in comment in code), the
parent app communication and local feeding state
management should probably be flipped in order to better
deal with possible failure in communication.

jasonzurita committed:
chore: move `AppState` to better group

jasonzurita committed:
Merge pull request #17 from jasonzurita/feature/higher-order-reducer

Higher order feeding communication reducer

jasonzurita committed:
chore: update todos

This appears to be working pretty well!

jasonzurita committed:
chore: clean up code style (minor)

jasonzurita committed:
chore: remove some outdated todos

jasonzurita committed:
feat: preform watch comms first then local state

This better address the _async_ nature of the
watch communication. We can now handle failure
communication cases better.

This also:
- Removes the messy higher order feeding communication reducer
- Sets up for receiving feeding state changes from the main app!

In the future, having the watch communication logic move out of the
individual views would be nice, but sending async actions is difficult
to handle right now.

jasonzurita committed:
fix: make sure feeding button UI is correct

Update needs to come after the actual start
of the feeding to make sure the UI reflects
the right state of the in-progress feeding.

jasonzurita committed:
feat: reflect actions from main app on apple watch

This will allow the apple watch to process
feeding actions. Still need to have the watch
_sync up_ with the main app when coming to life
for the first time, but this is already a good improvement.

jasonzurita committed:
Merge pull request #18 from jasonzurita/feature/handle-communication-from-main-app

Handle communication from main app

jasonzurita committed:
chore: move font files to common folder

to allow clearer reuse in watch app

jasonzurita committed:
feat: add custom fonts to watch app

They are now available to be used by name like:
`.font(Font.custom("NotoSans-BoldItalic", size: 30))`

Interestingly enough, if the font assets don't exist,
the system will default to using stock fonts. This
might be a good thing to have snapshot tests to lock down.

jasonzurita committed:
feat: update watch app fonts

The custom fonts extension makes for a nice
api for custom fonts while still preserving dynamic
type. Here is where that came from:
https://www.hackingwithswift.com/quick-start/swiftui/how-to-use-dynamic-type-with-a-custom-font

Also, update some todos

jasonzurita committed:
chore: clean up code (minor)

jasonzurita committed:
Merge pull request #19 from jasonzurita/feature/fonts

Add custom fonts to Apple Watch app

jasonzurita committed:
fix: stop main app communication

Unfortunately, when sending a feeding
from the watch to the phone, the communication
would "boomerang" back to the watch and duplicate
what was just sent (e.g., duplicate feedings would show
up on the watch). This commit breaks that communication
for the time being.

This was an oversight on my part :(. That
is okay though, this exposed how fragile
things were getting and that we should probably
be using the context for the main app's communication.

By using the context, that will also give the watch the
potential of receiving all the logged feedings!

jasonzurita committed:
Merge pull request #20 from jasonzurita/fix/communication

Stop main app communication to fix duplicate watch events

jasonzurita committed:
feat: add tests for main app Feeding model entity

This will help support refactoring to add
codable conformance.

Also, fixes the full paused time to make sure
we account for an ended feeding. If not, the paused
time could previously go past the ended time to
give an inflated amount of paused time, which is wrong. This wasn't an
issue previously because management of these properties was handled
correctly (i.e., lastPausedDate and endDate were mutually exclusive).

We should, however, also enforce in code that you can't set a
lastPausedDate at a later time than the endDate.

jasonzurita committed:
feat: move feeding + tests to common SPM

- Support for use in watch app
- Remove no longer used framework test target
- Need to move supply amount to support this move
- Remove Common example test

jasonzurita committed:
feat: update watch context with state of feedings

Using Codable, keep the watch connectivity context
up to date with all the feedings. This can now be used
to do several things:
- Have the watch react to changes made to feedings on
the main app
- Have the watch display all feeding history!

jasonzurita committed:
feat: keep watch feedings in sync with main app!

This means:
- Watch uses the codable feedings in the context to
do a mass update and replace the current feedings
state. This is fine since SwiftUI "just" reacts to the changes
and seamlessly keeps the UI moving smoothly and in sync with
what was done on the main app. SwiftUI really rocks!
- Introduce new reducer for mass updating of the feedings
due to an incoming context update
- Rename feeding reducers for better clarity
- Old feeding entity is no longer needed!

jasonzurita committed:
chore: remove unused constants

jasonzurita committed:
Merge pull request #21 from jasonzurita/feature/context-communication

Context communication (keep feedings in sync)

jasonzurita committed:
feat: clean up fonts integration

- Fix the failing to find the _regular_ font. There
was a naming issue. That is cleaned up now, so
the main app could also delete some _special_ code
that accounted for this.
- Add semi bold font to watch app
- move font assets to extension, which is clearer

jasonzurita committed:
feat: add custom color palette to watch

jasonzurita committed:
chore: syntax clean up (SwiftFormat)

jasonzurita committed:
chore: remove outdated todo

jasonzurita committed:
fix: remove duplicate newly started feeding

Due to the async nature of sending the communication
there is a chance that the watch will end up showing
two feedings when starting a feeding.
i.e., the context updating shows the new feeding
and then the watch can sometimes add a new feeding on
top of that if the response form the send message comes
in after.

jasonzurita committed:
chore: clarify todo

jasonzurita committed:
feat: improve synchronization of app state context

When starting the watch app, it will ask the main app for
the full app state in context form. The main app will then
send back the full state including feedings and logged in
status.

This greatly improves the initial starting up experience
of the watch app! Also, this sets the stage for the watch to
have a simple UI mechanism to refresh the full state from the
main app!!

jasonzurita committed:
feat: add placeholder text when no feedings

jasonzurita committed:
feat: remove new feeding actions / reducer

no longer needed since watch state is handled by
getting updates from the context. And, SwiftUI is great
at managing the UI updates due to context changes!

This also included the removal of the communication
failure state/action since that was a direct side effect
of the new feeding reducer.

jasonzurita committed:
Merge pull request #22 from jasonzurita/feature/improve-context-management

Improve synchronization of app state context (feedings and logged in state)

jasonzurita committed:
fix: restore saved fyi popup on feeding stop

Also, some renaming for better clarity

jasonzurita committed:
Merge pull request #23 from jasonzurita/feature/feeding-actions-cleanup

Feeding actions UI + cleanup

jasonzurita committed:
feat: abstract fyi dialog

Fyi dialog is now customizable!

This will also be used to show a watch <> phone
communication error

jasonzurita committed:
feat: show fyi dialog if send message fails

Leverages the now generic fyi dialog to show
the user that the watch <> phone communication
failed for a better UX.

Note: this dialog takes precedence over the saved
dialog, so that we can't awkwardly show both at the same time

jasonzurita committed:
Merge pull request #24 from jasonzurita/feature/surface-communication-error

Surface watch <> phone communication error

jasonzurita committed:
feat: implement loading spinner view

This will be used in loading situations like
when waiting for the main app to respond after
a trying to add a feeding from the watch.

The spinning animation starts on appear and
this view is meant to be thrown away like in an
if statement.

jasonzurita committed:
feat: make use of loading spinner

Show loading spinner on feeding actions
(start, pause, resume, stop).

The loading spinner will not be double shown
with dialogs since they are all mutually exclusive in
being shown due to the if / if else chain for showing popups
like this. Due to this, combining the dialog and loading booleans
into an enum might be a better way to better lock down the
mutual exclusivity.

jasonzurita committed:
Merge pull request #25 from jasonzurita/feature/loading-spinner

Implement loading spinner

jasonzurita committed:
fix: call correct end actions

I had them backwards! Weird UI/UX
behavior as a result.

jasonzurita committed:
feat: tweak dialog display times

to better be able to read the text on
the dialog popups

jasonzurita committed:
chore: update todos

jasonzurita committed:
feat: show error dialog if encoding fails

better UX

jasonzurita committed:
feat: add placeholder image when no feedings

jasonzurita committed:
chore: move color assets to watch extension

to be used in the extension code

jasonzurita committed:
feat: add extension for custom colors

to be used more easily and safely in code

e.g., Color.bpMediumBlue

jasonzurita committed:
feat: update colors to correct palette

jasonzurita committed:
feat: better style feeding timer chip

Puts type and side on the left of the timer.
This may change in the future, but a good first
pass

jasonzurita committed:
chore: clean up todos
